lib/portlayer/attach/communication/connector.go

```diff
@@ -112,33 +128,44 @@ func (c *Connector) Interaction(ctx context.Context, id string) (SessionInteract
 func (c *Connector) interaction(ctx context.Context, id string) (SessionInteractor, error) {
 	defer trace.End(trace.Begin(id))
 
-	conn := c.aliveAndKicking(ctx, id)
-	if conn != nil {
-		return conn, nil
-	}
-
 	if ctx.Err() == context.DeadlineExceeded {
 		return nil, fmt.Errorf("attach connector: no such connection")
 	}
 
+	conn, err := c.SessionIfAlive(ctx, id)
+	if conn != nil && err == nil {
+		return conn, nil
+	}
+
```

Comment:
should we move this above the context deadline exceeded check? It may reduce the edge case behaviour where we've got one foot on either side of the deadline (e.g. stdout returned before deadline, stderr hitting the timeout and failing). This is the case where singleflight doesn't block the calls because they're not quite concurrent.

The change I suggest should help reduce the frequency of the issue, however we should also be checking in the personality that all of the required streams are successful or failing - iirc we were previously able to get into a state where one would succeeded and we would have no stdin, for a container run with `-it` for example, but stdout was successfuly so the client wouldn't exit.